In [1]:
import logging
import numpy as np
import pandas as pd
root = logging.getLogger()
root.addHandler(logging.StreamHandler())
%matplotlib inline

In [2]:
# download http://bit.ly/1R8pt20 (zipped Turtles shapefiles), and unzip them
# or download from Google Drive: https://drive.google.com/open?id=0B9cazFzBtPuCSFp3YWE1V2JGdnc
from iSDM.species import IUCNSpecies
fish = IUCNSpecies(name_species='All')
fish.load_shapefile('../data/fish/FW_FISH.shp') # warning, 2GB of data will be loaded, may take a while!!

Enabled Shapely speedups for performance.
Loading data from: ../data/fish/FW_FISH.shp
The shapefile contains data on 12927 species areas.


## Simple data exploration

In [3]:
fish.get_data().columns

Index(['binomial', 'category', 'citation', 'class_name', 'compiler',
       'dist_comm', 'family_nam', 'genus_name', 'geometry', 'id_no', 'island',
       'kingdom_na', 'legend', 'order_name', 'origin', 'phylum_nam',
       'presence', 'seasonal', 'shape_area', 'shape_leng', 'source',
       'species_na', 'subpop', 'subspecies', 'tax_comm', 'year'],
      dtype='object')

How many unique binomials are there?

In [4]:
fish_data = fish.get_data()
fish_data['binomial'].unique().size 

6328

Get their names in an array:

In [5]:
unique_binomials = fish_data['binomial'].unique()
unique_binomials

array(['Bagarius suchus', 'Rasbora spilocerca', 'Notoglanidium pallidum',
       ..., 'Schistura procera', 'Monotrete cambodgiensis',
       'Acantopsis spectabilis'], dtype=object)

In [6]:
fish_data.head(10) # peak at the first 10 records

binomial category  \
0               Bagarius suchus       NT   
1            Rasbora spilocerca       LC   
2        Notoglanidium pallidum       VU   
3  Parananochromis brevirostris       VU   
4  Parananochromis brevirostris       VU   
5  Parananochromis brevirostris       VU   
6            Ambassis vachellii       LC   
7          Parasikukia maculata       LC   
8          Parasikukia maculata       LC   
9          Parasikukia maculata       LC   

                                            citation      class_name  \
0             Indo-Burma freshwater assessment, IUCN  ACTINOPTERYGII   
1             Indo-Burma freshwater assessment, IUCN  ACTINOPTERYGII   
2  IUCN (International Union for Conservation of ...  ACTINOPTERYGII   
3  IUCN (International Union for Conservation of ...  ACTINOPTERYGII   
4  IUCN (International Union for Conservation of ...  ACTINOPTERYGII   
5                                               None  ACTINOPTERYGII   
6  H.Larson and IUCN (International Union for Con...  ACTINOPTERYGII   
7             Indo-Burma freshwater assessment, IUCN  ACTINOPTERYGII   
8             Indo-Burma freshwater assessment, IUCN  ACTINOPTERYGII   
9             Indo-Burma freshwater assessment, IUCN  ACTINOPTERYGII   

          compiler dist_comm   family_nam       genus_name  \
0         IUCN FBU      None    SISORIDAE         Bagarius   
1   J. Scott, IUCN      None   CYPRINIDAE          Rasbora   
2             None      None  CLAROTEIDAE    Notoglanidium   
3             None      None    CICHLIDAE  Parananochromis   
4             None      None    CICHLIDAE  Parananochromis   
5             None      None    CICHLIDAE  Parananochromis   
6  M.Raiwalui/IUCN      None   AMBASSIDAE         Ambassis   
7   D. Allen, IUCN      None   CYPRINIDAE      Parasikukia   
8   D. Allen, IUCN      None   CYPRINIDAE      Parasikukia   
9         IUCN FBU      None   CYPRINIDAE      Parasikukia   

                                            geometry     id_no   ...    \
0  POLYGON ((101.8125000000001 22.48333333300008,...  181271.0   ...     
1  POLYGON ((103.5121466740001 18.53992445200004,...  181127.0   ...     
2  POLYGON ((12.81128506100004 -4.286932711999953...  182359.0   ...     
3  (POLYGON ((12.10738288000005 1.628413222000063...  182240.0   ...     
4  (POLYGON ((12.78953162300007 2.117309570000032...  182240.0   ...     
5  (POLYGON ((13.13924221400003 -0.98298000799997...  182240.0   ...     
6  (POLYGON ((105.2625 -6.754166666999936, 105.25...  166893.0   ...     
7  (POLYGON ((100.374432712 18.59657999700005, 10...  181017.0   ...     
8  (POLYGON ((100.8458333330001 12.69166666700005...  181017.0   ...     
9  (POLYGON ((99.65793745300005 16.53653225300008...  181017.0   ...     

  presence seasonal  shape_area   shape_leng                           source  \
0      2.0      5.0   49.374794    60.288418              Red List assessment   
1      2.0      5.0    8.361239    58.345301              Red List assessment   
2      1.0      1.0    1.524058    10.000992                             None   
3      2.0      1.0    5.563441    30.269691                             None   
4      1.0      1.0    4.872947    39.634665                             None   
5      1.0      1.0    0.366030     6.770438                             None   
6      1.0      1.0  121.742461  1817.507950                         H.Larson   
7      2.0      1.0   12.017594    58.052434              Red List assessment   
8      2.0      1.0    4.081329    37.348873  C. Vidthayanon pers. comm. 2011   
9      2.0      5.0    0.374677     6.355700              Red List assessment   

     species_na  subpop  subspecies  tax_comm    year  
0        suchus    None        None      None  2012.0  
1    spilocerca    None        None      None  2012.0  
2      pallidum    None        None      None     0.0  
3  brevirostris    None        None      None     0.0  
4  brevirostris    None        None      None     0.0  
5  brevirostris    

Sort them by their "shape_area" column, to get the biggest and smallest areas

In [7]:
fish_data.sort(columns='shape_area', inplace=True)

/home/daniela/anaconda/envs/biodiversity_py3/lib/python3.4/site-packages/ipykernel/__main__.py:1: FutureWarning: sort(columns=....) is deprecated, use sort_values(by=.....)
  if __name__ == '__main__':


Peak at the first 5 records with the smallest area

In [8]:
fish_data.head(5)

binomial category  \
1396   Astatotilapia burtoni       LC   
6610          Channa striata       LC   
6889         Rasbora rasbora       LC   
11988          Puntius chola       LC   
5382       Pisodonophis boro       LC   

                                                citation      class_name  \
1396   IUCN (International Union for Conservation of ...  ACTINOPTERYGII   
6610                                   Chaudhry S., 2009  ACTINOPTERYGII   
6889                                                IUCN  ACTINOPTERYGII   
11988                                      Dahanukar, N.  ACTINOPTERYGII   
5382                                   Chaudhry S., 2009  ACTINOPTERYGII   

               compiler dist_comm    family_nam     genus_name  \
1396               None      None     CICHLIDAE  Astatotilapia   
6610   Shivaji Chaudhry      None     CHANNIDAE         Channa   
6889      W. Vishwanath      None    CYPRINIDAE        Rasbora   
11988     Dahanukar, N.      None    CYPRINIDAE        Puntius   
5382   Shivaji Chaudhry      None  OPHICHTHIDAE   Pisodonophis   

                                                geometry     id_no   ...    \
1396   POLYGON ((29.81538226200007 -5.175785492999978...   60462.0   ...     
6610   POLYGON ((121.2046038030001 14.44001242700006,...  166563.0   ...     
6889   POLYGON ((97.65833333300009 16.53333333300003,...  166440.0   ...     
11988  POLYGON ((97.65833333300009 16.53333333300003,...  166443.0   ...     
5382   POLYGON ((85.46407583500007 19.66257255000005,...  166552.0   ...     

      presence seasonal shape_area shape_leng  \
1396       1.0      0.0   0.000122   0.060842   
6610       1.0      4.0   0.000242   0.073176   
6889       1.0      1.0   0.000286   0.075311   
11988      1.0      1.0   0.000286   0.075311   
5382       1.0      4.0   0.000393   0.101212   

                                                  source species_na  subpop  \
1396                                                None    burtoni    None   
6610                                          CAS 204572    striata    None   
6889                                      Fishbase, 2026    rasbora    None   
11988  accessed through GBIF data portal, NRM-Fishes,...      chola    None   
5382                                         FISH 387621       boro    None   

       subspecies  tax_comm    year  
1396         None      None     0.0  
6610         None      None  2007.0  
6889         None      None  2007.0  
11988        None      None  2007.0  
5382         None      None  2007.0  

[5 rows x 26 columns]

The last 5 records (biggest area)

In [9]:
fish_data.tail(5)

binomial category                            citation  \
10436    Rutilus rutilus       LC                                None   
271      Cottus confusus       LC  Freshwater Biodiversity Unit, IUCN   
4886   Perca fluviatilis       LC                                None   
5973           Lota lota       LC                                None   
10370        Esox lucius       LC                                None   

           class_name                       compiler dist_comm  family_nam  \
10436  ACTINOPTERYGII                           None      None  CYPRINIDAE   
271    ACTINOPTERYGII  Phillipa Palmeirin, FBU, IUCN      None    COTTIDAE   
4886   ACTINOPTERYGII                           None      None    PERCIDAE   
5973   ACTINOPTERYGII                           None      None     LOTIDAE   
10370  ACTINOPTERYGII                           None      None    ESOCIDAE   

      genus_name                                           geometry     id_no  \
10436    Rutilus  (POLYGON ((-4.770833332999928 48.5250000000000...   19787.0   
271       Cottus  (POLYGON ((-167.666666666 65.79166666700007, -...  202659.0   
4886       Perca  (POLYGON ((-4.770833332999928 48.5250000000000...   16580.0   
5973        Lota  (POLYGON ((4.038374837000049 51.96308458100003...  135675.0   
10370       Esox  (POLYGON ((-4.770833332999928 48.5250000000000...  135631.0   

        ...   presence seasonal   shape_area   shape_leng  \
10436   ...        1.0      1.0  2987.465593  1978.210580   
271     ...        2.0      1.0  3031.001694  5133.731779   
4886    ...        1.0      1.0  3365.103170  2218.467787   
5973    ...        1.0      1.0  4264.651654  4250.283811   
10370   ...        1.0      1.0  5225.451235  4405.111110   

                                                  source   species_na  subpop  \
10436                                               None      rutilus    None   
271    Page, L.M. and Burr, B.M. 2011. Peterson Field...     confusus    None   
4886                                                None  fluviatilis    None   
5973                                                None         lota    None   
10370                                               None       lucius    None   

       subspecies  tax_comm    year  
10436        None      None     0.0  
271          None      None  2012.0  
4886         None      None     0.0  
5973         None      None     0.0  
10370        None      None     0.0  

[5 rows x 26 columns]

Decision: **"We will include all species except those that are completely extinct. For the (possibly) extant species, we will include their entire range (including polygons from which they are currently extinct) in order to cover their full potential distribution."**

From IUCN, these are the codes for presence:
1. Extant
2. Probably extant
3. Possibly extant
4. Possibly extinct
5. Extinct
6. Presence uncertain

Are there species for which the presence category isn't any of the above? (the "~" is negation, so this below says "select all fish data, for which the "presence" column is NOT in [1,2,3,4,5,6])

In [10]:
fish_data[~fish_data.presence.isin([1,2,3,4,5,6])]

binomial category  \
3269            Acipenser fulvescens       LC   
2312            Sicyopterus sarasini       EN   
4669     Chanodichthys erythropterus       LC   
11591         Hemiculter leucisculus       LC   
3292   Microphysogobio tungtingensis       NT   
12321            Carassius carassius       LC   
10823                    Tinca tinca       LC   
1911      Rhynchocypris czekanowskii       LC   
6126        Chanodichthys mongolicus       LC   
1054                Acipenser baerii       EN   
1139             Coregonus pidschian       LC   
3270            Acipenser fulvescens       LC   
2349             Agosia chrysogaster       LC   
10621                 Leuciscus idus       LC   
10718           Coregonus autumnalis       LC   
9931            Cyprinus rubrofuscus       LC   
12919               Rhodeus sericeus    LR/lc   
10367                    Esox lucius       LC   
9015      Misgurnus anguillicaudatus       LC   
8574                  Silurus asotus       LC   
4163             Pseudorasbora parva       LC   
4887               Perca fluviatilis       LC   
1086             Cobitis melanoleuca       LC   
9419         Rhynchocypris percnurus       LC   
10525             Thymallus arcticus       LC   
5967                       Lota lota       LC   
3271            Acipenser fulvescens       LC   
3490              Ambassis urotaenia       LC   

                                 citation      class_name  \
3269   Freshwater Biodiversity Unit, IUCN  ACTINOPTERYGII   
2312                                 None  ACTINOPTERYGII   
4669                                 None  ACTINOPTERYGII   
11591                                None  ACTINOPTERYGII   
3292                                 None  ACTINOPTERYGII   
12321                                None  ACTINOPTERYGII   
10823                                None  ACTINOPTERYGII   
1911                                 None  ACTINOPTERYGII   
6126                                 None  ACTINOPTERYGII   
1054                                 None  ACTINOPTERYGII   
1139                                 None  ACTINOPTERYGII   
3270                          NatureServe  ACTINOPTERYGII   
2349                          NatureServe  ACTINOPTERYGII   
10621                                None  ACTINOPTERYGII   
10718                                None  ACTINOPTERYGII   
9931                                 None  ACTINOPTERYGII   
12919                                None  ACTINOPTERYGII   
10367                                None  ACTINOPTERYGII   
9015                                 None  ACTINOPTERYGII   
8574                                 None  ACTINOPTERYGII   
4163                                 None  ACTINOPTERYGII   
4887                                 None  ACTINOPTERYGII   
1086                                 None  ACTINOPTERYGII   
9419                                 None  ACTINOPTERYGII   
10525                                None  ACTINOPTERYGII   
5967                                 None  ACTINOPTERYGII   
3271                                 None  ACTINOPTERYGII   
3490                                 None  ACTINOPTERYGII   

                                 compiler dist_comm     family_nam  \
3269                                 IUCN      None  ACIPENSERIDAE   
2312                                 None      None       GOBIIDAE   
4669                                 None      None     CYPRINIDAE   
11591                                None      None     CYPRINIDAE   
3292                                 None      None     CYPRINIDAE   
12321                                None      None     CYPRINIDAE   
10823                                None      None     CYPRINIDAE   
1911                                 None      None     CYPRINIDAE   
6126                                 None      None     CYPRINIDAE   
1054                                 None      None  ACIPENSERIDAE   
1139                                 None      None     SALMONIDAE

Apparently there are 28 species for which the "presence" is set to 0 in some regions. Their binomials are:

In [11]:
fish_data[~fish_data.presence.isin([1,2,3,4,5,6])]['binomial']

3269              Acipenser fulvescens
2312              Sicyopterus sarasini
4669       Chanodichthys erythropterus
11591           Hemiculter leucisculus
3292     Microphysogobio tungtingensis
12321              Carassius carassius
10823                      Tinca tinca
1911        Rhynchocypris czekanowskii
6126          Chanodichthys mongolicus
1054                  Acipenser baerii
1139               Coregonus pidschian
3270              Acipenser fulvescens
2349               Agosia chrysogaster
10621                   Leuciscus idus
10718             Coregonus autumnalis
9931              Cyprinus rubrofuscus
12919                 Rhodeus sericeus
10367                      Esox lucius
9015        Misgurnus anguillicaudatus
8574                    Silurus asotus
4163               Pseudorasbora parva
4887                 Perca fluviatilis
1086               Cobitis melanoleuca
9419           Rhynchocypris percnurus
10525               Thymallus arcticus
5967                     

Actually it's 26, since "Acipenser fulvescens" has 3 separate records (polygons)

"For the (possibly) extant species, we will include their entire range (including polygons from which they are 
currently extinct)"

As an example, here is a species with both extant and extinct areas

In [12]:
fish_data[fish_data.binomial=='Acantharchus pomotis']

binomial category     citation      class_name  \
2154  Acantharchus pomotis       LC  NatureServe  ACTINOPTERYGII   
2156  Acantharchus pomotis       LC         None  ACTINOPTERYGII   
2155  Acantharchus pomotis       LC  NatureServe  ACTINOPTERYGII   

                                compiler dist_comm     family_nam  \
2154  Freshwater Biodiversity Unit, IUCN      None  CENTRARCHIDAE   
2156                                None      None  CENTRARCHIDAE   
2155  Freshwater Biodiversity Unit, IUCN      None  CENTRARCHIDAE   

        genus_name                                           geometry  \
2154  Acantharchus  POLYGON ((-83.63333333299994 31.23750000000007...   
2156  Acantharchus  (POLYGON ((-83.94131181499995 32.4517130540000...   
2155  Acantharchus  (POLYGON ((-84.33333333299998 30.0500000000000...   

         id_no   ...   presence seasonal shape_area  shape_leng  \
2154  201938.0   ...        5.0      0.0   0.024246    0.796963   
2156  201938.0   ...        1.0      0.0   2.006039   18.544973   
2155  201938.0   ...        1.0      0.0  25.110455  170.717198   

                                                 source species_na  subpop  \
2154  Digital Distribution Maps of the Freshwater Fi...    pomotis    None   
2156                                               None    pomotis    None   
2155  Digital Distribution Maps of the Freshwater Fi...    pomotis    None   

      subspecies  tax_comm    year  
2154        None      None  2013.0  
2156        None      None     0.0  
2155        None      None  2013.0  

[3 rows x 26 columns]

Plenty of columns, let's just select the 'binomial', 'presence', 'geometry', and 'shape_area' to have a clearer overview

In [13]:
fish_data[fish_data.binomial=='Acantharchus pomotis'][['binomial','presence', 'geometry','shape_area']]

binomial  presence  \
2154  Acantharchus pomotis       5.0   
2156  Acantharchus pomotis       1.0   
2155  Acantharchus pomotis       1.0   

                                               geometry  shape_area  
2154  POLYGON ((-83.63333333299994 31.23750000000007...    0.024246  
2156  (POLYGON ((-83.94131181499995 32.4517130540000...    2.006039  
2155  (POLYGON ((-84.33333333299998 30.0500000000000...   25.110455

As agreed, we want to keep the first (extinct==5) polygon when there are other regions (the next two records) where the species is not extinct. There is a function that can do the filtering out in a simple way.

In [14]:
fish.drop_extinct_species()

There are currently 6328 unique species. 

Filtering out the following extinct species: ['Alburnus akili', 'Alburnus nicaeensis', 'Aphanius splendens', "Aplocheilichthys sp. nov. 'Naivasha'", 'Barbus microbarbis', 'Catostomus bernardini', 'Chasmistes muriei', 'Coregonus bezola', 'Coregonus fera', 'Coregonus gutturosus', 'Coregonus hiemalis', 'Coregonus johannae', 'Coregonus oxyrinchus', 'Coregonus restrictus', 'Cottus bendirei', 'Cottus echinatus', 'Cyprinodon arcuatus', 'Cyprinus micristius', 'Empetrichthys latos', 'Empetrichthys merriami', "Eudontomyzon sp. nov. 'migratory'", 'Fundulus albolineatus', 'Gambusia amistadensis', 'Gambusia georgei', 'Gambusia senilis', 'Gasterosteus crenobiontus', 'Gila crassicauda', 'Ictalurus pricei', 'Lepidomeda altivelis', 'Mirogrex hulensis', 'Moxostoma lacerum', 'Noturus trautmani', 'Pantanodon madagascariensis', 'Platytropius siamensis', 'Pogonichthys ciscoides', 'Prototroctes oxyrhynchus', 'Ptychochromis onilahy', 'Ptychochromoides itasy', 'Rhinic

In [15]:
fish.save_data("../data/fish/selection/non_extinct.pkl", method="pickle")

Saved data: ../data/fish/selection/non_extinct.pkl 
Type of data: <class 'geopandas.geodataframe.GeoDataFrame'> 


In [16]:
non_extinct_fish = fish.get_data()

### Cross-check with GBIF records

#### How many freshwater fish observation records do we have (without any filtering)


In [17]:
total_records = 0
# download the file "merged.msg" from here: (My Google Drive)
# https://drive.google.com/open?id=0B9cazFzBtPuCTHcyTmVXV0pNT0k
# and place it the subfolder "selection"
# The "merged.msg" file contains all datagrames for all individual species.
for df in pd.read_msgpack("../data/fish/selection/merged.msg", iterator=True):
    total_records += df.shape[0]    

In [18]:
total_records

4733792

#### 1. Problematic/unavailable locality
Which species have **no latitude/longitude information, and how many "observations" are there for them?** What is the basis of record for such species? (Suspicion is that these are old museum/journals records)

In [19]:
important_columns = ['species', 'dateidentified','basisofrecord', 'verbatimlocality', 'day', 'month', 'year' ]
result_no_lat_long = pd.DataFrame(columns=important_columns)

for df in pd.read_msgpack("../data/fish/selection/merged.msg", iterator=True):
    if "decimallatitude" not in df.columns.tolist() or "decimallongitude" not in df.columns.tolist():
        common_columns = list(set(important_columns).intersection(set(df.columns.tolist())))
        result_no_lat_long = result_no_lat_long.append(df[common_columns], ignore_index=True)
      

#### How many unique species are there, without lat/lon?

In [20]:
result_no_lat_long['species'].unique().size

600

Warning: this does not mean that the species ONLY has such records. It could be that particular species also has records WITH latitude/longitude information.

##### So then what percentage of the total number of records?

In [21]:
(result_no_lat_long.shape[0]/total_records) * 100 # not that bad, only about 0.2%

0.16261804489931117

#### What about verbatim locality? Is that helpful? These records have a verbatim locality

In [22]:
result_no_lat_long[['species', 'verbatimlocality']][result_no_lat_long.verbatimlocality.notnull()]

species  \
27             Salmo stomachicus   
42       Clypeobarbus bellcrossi   
114      Squalidus atromaculatus   
869           Salvelinus perisii   
1285            Barbus fritschii   
1290            Barbus fritschii   
1292            Barbus fritschii   
1293            Barbus fritschii   
1341                  Hara horai   
1367            Barbus callensis   
1368            Barbus callensis   
1599         Coregonus vandesius   
1708        Barbodes wynaadensis   
2066            Barbus callensis   
2113          Nannocharax wittei   
3247                Barbus nasus   
3248                Barbus nasus   
3362  Luciobarbus brachycephalus   
3369  Luciobarbus brachycephalus   
3446         Triplophysa grahami   
3465              Barbus petenyi   
3466              Barbus petenyi   
3467              Barbus petenyi   
3666              Barbus petenyi   
4095       Sabanejewia bulgarica   
4205          Coregonus nilssoni   
4209          Coregonus nilssoni   
4375      Haplochromis xenostoma   
4603        Haplochromis bartoni   
4820      Sabanejewia vallachica   
4823      Sabanejewia vallachica   
6480        Devario kakhienensis   
6514              Devario browni   
6515              Devario browni   
6516              Devario browni   
6517              Devario browni   
6518              Devario browni   
6519              Devario browni   
6520              Devario browni   
6525              Devario browni   
6530              Devario browni   
6579           Chondrostoma cyri   
6862            Barbus callensis   
6872            Barbus callensis   
6877            Barbus callensis   
6885            Barbus callensis   
6889            Barbus callensis   
7115            Barbus callensis   
7116            Barbus callensis   
7117            Barbus callensis   
7415            Synodontis omias   
7449         Pollimyrus guttatus   

                                       verbatimlocality  
27                                        ; ; ; ; ; ; ;  
42                   Africa | Zambia | Western Province  
114                                              Hainan  
869                                       ; ; ; ; ; ; ;  
1285                                            Berkana  
1290                                       Oued Tensift  
1292                                            Oued Za  
1293                                         Ras-el-Ain  
1341                                        unspecified  
1367                                       Oued Tensift  
1368                                       Oued Tensift  
1599  ; [Scotland; U.K.]; ; ; ; ; ; Lochmaben; Dumfr...  
1708                         Vithry [India or Pakistan]  
2066                                       Oued Tensift  
2113            Between Kermani and Kasai [Belge Congo]  
3247                                                Fes  
3248                                                Fes  
3362  Eurasia | Union Of Soviet Socialist Republics ...  
3369                      Leukoran; Rec'd 24 Dec. 1885.  
3446                     ; W. China; ; ; ; ; ; Yunnanfu  
3465         Milcov R. at Focsani, S Moldavia;  Vrancea  
3466  Cerna R. at Baile-Herculane, Banat;  Caras-Sev...  
3467  Hirtibaciu R., indirect trib. to Olt R. at Moh...  
3666                                            Croatia  
4095  Confluence of Arges R. with Danube R. at Olten...  
4205                                  Lake Ring, Scania  
4209                                  Lake Ring, Scania  
4375             Lake Victoria [exact locality unknown]  
4603    Aquarium stock; originating from Lake Victoria.  
4820         Milcov R. at Focsani, S Moldavia;  Vrancea  
4823     Ialomita R. at Crivina, N of Bucharest;  Ilfov  
6480                                        Phagna Dam.  
6514  ~25 km N Mae Hong Son town, on road to Chiang ...  
6515  Mae Sariang district, 20 km SE of  Mae Sariang...  
6516  ~8 km SW Mae Hong Son town, Huey Sertao, (old ...  
6517  village of Mae Sam Lae

How many records with verbatim locality?

In [23]:
result_no_lat_long[['species', 'verbatimlocality']][result_no_lat_long.verbatimlocality.notnull()].shape[0]

52

#### If we group them by the basisofrecord, how many?

In [24]:
grouped_no_lat_lon = pd.DataFrame()
grouped_no_lat_lon['count'] = result_no_lat_long.groupby(['species', 'basisofrecord']).apply(lambda x: x['basisofrecord'].count())
grouped_no_lat_lon.head(30)

count
species                    basisofrecord            
Abactochromis labrosus     PRESERVED_SPECIMEN      6
Acanthobrama microlepis    PRESERVED_SPECIMEN     12
Acanthocobitis pavonacea   PRESERVED_SPECIMEN      1
Acentrogobius therezieni   PRESERVED_SPECIMEN      9
Acheilognathus deignani    PRESERVED_SPECIMEN      2
Acheilognathus meridianus  PRESERVED_SPECIMEN      1
Acrochordonichthys gyrinus PRESERVED_SPECIMEN      2
Acrossocheilus microstoma  PRESERVED_SPECIMEN      1
Ailiichthys punctata       PRESERVED_SPECIMEN      4
Akysis manipurensis        PRESERVED_SPECIMEN      4
Akysis pictus              PRESERVED_SPECIMEN      4
Akysis portellus           PRESERVED_SPECIMEN      1
Alburnoides eichwaldii     PRESERVED_SPECIMEN     18
Alburnoides fasciatus      PRESERVED_SPECIMEN     14
Alburnoides gmelini        PRESERVED_SPECIMEN      4
Alburnoides kubanicus      PRESERVED_SPECIMEN      5
Alburnoides maculatus      PRESERVED_SPECIMEN      8
Alburnoides rossicus       PRESERVED_SPECIMEN     23
Alburnus escherichii       PRESERVED_SPECIMEN      4
Alburnus filippii          PRESERVED_SPECIMEN     38
Alburnus tarichi           PRESERVED_SPECIMEN      4
Alosa killarnensis         PRESERVED_SPECIMEN      1
Ambastaia nigrolineata     OBSERVATION             1
                           PRESERVED_SPECIMEN      1
Ambastaia sidthimunki      PRESERVED_SPECIMEN      1
Amblyceps carinatum        PRESERVED_SPECIMEN      1
Amblyceps protentum        PRESERVED_SPECIMEN      1
Anabarilius alburnops      PRESERVED_SPECIMEN      6
Aphyosemion abacinum       PRESERVED_SPECIMEN      9
Aphyosemion alpha          PRESERVED_SPECIMEN      5

#### Which ones have at least 50 records (more likely we need at least 100 records in total per species, but let's keep the bar low still)

In [25]:
grouped_no_lat_lon[grouped_no_lat_lon['count']>50]

count
species                   basisofrecord            
Aphyosemion mimbon        PRESERVED_SPECIMEN     61
Barbus callensis          PRESERVED_SPECIMEN    122
Barbus ciscaucasicus      PRESERVED_SPECIMEN    673
Barbus fritschii          PRESERVED_SPECIMEN     95
Barbus petenyi            PRESERVED_SPECIMEN    553
Chondrostoma oxyrhynchum  PRESERVED_SPECIMEN     72
Cobitis stephanidisi      PRESERVED_SPECIMEN     62
Coregonus nilssoni        PRESERVED_SPECIMEN     59
Distichodus nefasch       PRESERVED_SPECIMEN     54
Haplochromis bareli       PRESERVED_SPECIMEN    117
Haplochromis crocopeplus  PRESERVED_SPECIMEN     65
Haplochromis dentex       PRESERVED_SPECIMEN    186
Haplochromis heusinkveldi PRESERVED_SPECIMEN   1062
Haplochromis mento        PRESERVED_SPECIMEN     63
Haplochromis michaeli     PRESERVED_SPECIMEN    173
Haplochromis perrieri     PRESERVED_SPECIMEN    340
Haplochromis pyrrhopteryx PRESERVED_SPECIMEN     54
Haplochromis tanaos       PRESERVED_SPECIMEN    140
Haplochromis xenostoma    PRESERVED_SPECIMEN     53
Oxynoemacheilus brandtii  PRESERVED_SPECIMEN     52

According to http://gbif.github.io/gbif-api/apidocs/org/gbif/api/vocabulary/BasisOfRecord.html
these are different from "real" human observations. But I'm not sure what PRESERVED_SPECIMEN is exactly?

#### What about the year in which these observations have been recorded? For those after 1960... (note that not all records have a year or any date information)

In [26]:
result_no_lat_long_years = result_no_lat_long[['species','year']][result_no_lat_long.year > 1960].groupby('species')['year'].apply(lambda x:x.tolist())
pd.DataFrame(result_no_lat_long_years).head(30)

year
species                                                                         
Acanthobrama microlepis                 [2006.0, 2006.0, 1966.0, 1966.0, 1966.0]
Acentrogobius therezieni                                [1963.0, 1963.0, 1963.0]
Acrochordonichthys gyrinus                                      [1995.0, 1995.0]
Alburnoides eichwaldii                  [2006.0, 2006.0, 2006.0, 2006.0, 2006.0]
Alburnoides fasciatus          [2010.0, 2010.0, 2007.0, 2007.0, 2007.0, 2006....
Alburnoides gmelini                             [2004.0, 2004.0, 2004.0, 2004.0]
Alburnoides rossicus                            [2012.0, 2012.0, 2012.0, 2012.0]
Alburnus escherichii                            [2006.0, 2006.0, 1990.0, 1990.0]
Alburnus filippii                                       [2012.0, 2010.0, 2006.0]
Ambastaia nigrolineata                                                  [2011.0]
Ambastaia sidthimunki                                                   [2007.0]
Amblyceps protentum                                                     [2012.0]
Aphyosemion abacinum                                            [1976.0, 1976.0]
Aphyosemion alpha                               [1994.0, 1993.0, 1993.0, 1993.0]
Aphyosemion chauchei                                            [1978.0, 1978.0]
Aphyosemion congicum                                                    [1971.0]
Aphyosemion exigoideum                                          [1977.0, 1976.0]
Aphyosemion mimbon                                                      [1976.0]
Aphyosemion seegersi                                            [1978.0, 1978.0]
Aplocheilichthys antinorii                                              [1982.0]
Arius madagascariensis          [2002.0, 1988.0, 1988.0, 1964.0, 1962.0, 1962.0]
Astatoreochromis vanderhorsti           [1992.0, 1992.0, 1992.0, 1991.0, 1983.0]
Astatotilapia flaviijosephi                                     [2002.0, 1972.0]
Aulonocara ethelwynnae                                          [1985.0, 1985.0]
Aulonocara hansbaenschi                                         [1985.0, 1985.0]
Aulonocara korneliae                                            [1985.0, 1985.0]
Aulonocara maylandi                                     [1985.0, 1984.0, 1980.0]
Aulonocara steveni                                              [1985.0, 1985.0]
Aulonocara stonemani                            [1973.0, 1971.0, 1971.0, 1971.0]
Barbus aliciae                                                          [1980.0]

In [ ]:
pd.DataFrame(result_no_lat_long_years).shape[0]

228

228 unique species with records after 1960, but no useful locality info about them anyways. I think all records without lat/lon will need to be discarded, for the 600 species.

Further selection criteria...

#### 2. Records with latitude/longitude: how many? how many with year>1960? year>1990? Without any date? accuracy of location?
warning: below takes long to run, processes occurrences for more than 5000 unique species

In [ ]:
import pandas as pd
important_columns1 = ['species', 'dateidentified', 'eventdate', 'basisofrecord', 'decimallatitude','decimallongitude', 'day', 'month', 'year' ]
result_with_lat_long = pd.DataFrame(columns=important_columns1)
counter = 0
for df in pd.read_msgpack("../data/fish/selection/merged.msg", iterator=True):
    counter += 1
    if (counter%100==0):
        print("Processing.. ", counter)
    if "decimallatitude" in df.columns.tolist() and "decimallongitude" in df.columns.tolist():
        common_columns = list(set(important_columns1).intersection(set(df.columns.tolist())))
        result_with_lat_long = result_with_lat_long.append(df[common_columns], ignore_index=True)
      

Processing..  100
Processing..  200
Processing..  300
Processing..  400
Processing..  500
Processing..  600
Processing..  700
Processing..  800
Processing..  900
Processing..  1000
Processing..  1100
Processing..  1200
Processing..  1300
Processing..  1400
Processing..  1500
Processing..  1600
Processing..  1700
Processing..  1800
Processing..  1900
Processing..  2000


In [ ]:
result_with_lat_long = result_with_lat_long[result_with_lat_long.decimallatitude.notnull() & result_with_lat_long.decimallongitude.notnull()]

In [ ]:
result_with_lat_long.shape[0] # this-many occurrence records

In [ ]:
result_with_lat_long.shape[0]/ total_records * 100 # percentage of records out of *ALL* species records

### How many unique species have occurrence records with latitude/longitude?

In [ ]:
result_with_lat_long['species'].unique().size

### What percentage of the occurrence records HAVE latitude/longitude, but NO event date?

In [ ]:
result_with_lat_long_no_date = result_with_lat_long[(result_with_lat_long.eventdate.isnull()) & (result_with_lat_long.year.isnull())]

In [ ]:
result_with_lat_long_no_date.shape[0]/result_with_lat_long.shape[0] * 100 # 12% 

Best to take into account all observations which have either "year" or "eventdate" present. (or both)
Let's group them by species name, and count the number of observation records.

In [ ]:
grouped_lat_long_year_or_eventdate = pd.DataFrame()
grouped_lat_long_year_or_eventdate['count'] = result_with_lat_long[result_with_lat_long.eventdate.notnull() | result_with_lat_long.year.notnull()].groupby(['species']).apply(lambda x: x['species'].count())
grouped_lat_long_year_or_eventdate.head(10) # peak at the top 10 only

Notice that some of them have very few observations. We will filter them out later.
### How many unique species HAVE records with latitude/longitude, AND date of event (at least year)

In [ ]:
grouped_lat_long_year_or_eventdate.shape[0]

#### What percentage of observations is that?

In [ ]:
100 * result_with_lat_long[result_with_lat_long.eventdate.notnull() | result_with_lat_long.year.notnull()].shape[0] / result_with_lat_long.shape[0]

### How many unique species with latitude/longitude, AND event date after 1960?

In [ ]:
year_or_eventdate_1960 = result_with_lat_long[['species', 'year', 'eventdate']][(result_with_lat_long.year>1960) | (result_with_lat_long.eventdate>"1960")]

grouped_year_or_eventdate_1960 = pd.DataFrame()
grouped_year_or_eventdate_1960['numobservations'] = year_or_eventdate_1960.groupby(['species']).apply(lambda x: x['species'].count())
grouped_year_or_eventdate_1960.shape[0]

#### What percentage of observations is that?  (out of all records with latitude/longitude)?

In [ ]:
year_or_eventdate_1960.shape[0]/result_with_lat_long.shape[0] * 100

### From these above, how many species have 50+ observations?

In [ ]:
grouped_year_or_eventdate_1960[grouped_year_or_eventdate_1960.numobservations>=50]

Not that many. 1283 species.

#### What percentage is that (out of all records with latitude/longitude)?

In [ ]:
(grouped_year_or_eventdate_1960[grouped_year_or_eventdate_1960.numobservations>50].numobservations.sum() / result_with_lat_long.shape[0]) * 100

So 80% of all records with latitude and longitude, **are** with year>1960 **and** more than 50 occurrences per species.
But  **note that this still covers only 1281 uniqute species (out of 6280 that we started with) **

#### What percentage is that (out of all records with latitude/longitude AND year>1960)?

In [ ]:
100 * grouped_year_or_eventdate_1960[grouped_year_or_eventdate_1960.numobservations>50].numobservations.sum() / grouped_year_or_eventdate_1960.numobservations.sum()

Ok that means most (98%) species with latitude/longitude after year 1960, **have** more than 50 records.

### If we limit to records >1990, it becomes even more worrying.

In [ ]:
year_or_eventdate_1990 = result_with_lat_long[['species', 'year', 'eventdate', 'basisofrecord']][(result_with_lat_long.year>=1990) | (result_with_lat_long.eventdate>="1990")]

grouped_year_or_eventdate_1990 = pd.DataFrame()
grouped_year_or_eventdate_1990['numobservations'] = year_or_eventdate_1990.groupby(['species']).apply(lambda x: x['species'].count())
grouped_year_or_eventdate_1990.shape[0] # number of unique species

#### What percentage of observations is that?  (out of all records with latitude/longitude)?

In [ ]:
year_or_eventdate_1990.shape[0]/result_with_lat_long.shape[0] * 100

### From these, how many species have 50+ observations?

In [ ]:
grouped_year_or_eventdate_1990[grouped_year_or_eventdate_1990.numobservations>=50]

## Select a representative set of species for testing workflow
From meeting: ***The test set should include at least 1) the species for which lab test data on thermal tolerance are available and 2) species with very small and very large ranges, in order to find out whether the point sampling and modelling also works for these extremes.***

The non_extinct_fish dataframe is sorted in ascending order, according to the area size. Take the last binomial.

### Largest area

In [ ]:
non_extinct_fish.tail(1).binomial

See all the records(areas) for this binomial

In [ ]:
non_extinct_fish[non_extinct_fish.binomial=="Esox lucius"]

In [ ]:
esox_lucius = IUCNSpecies(name_species="Esox lucius")

In [ ]:
esox_lucius.set_data(non_extinct_fish[non_extinct_fish.binomial=="Esox lucius"])

In [ ]:
esox_lucius.save_shapefile("../data/fish/selection/esox_lucius")

In [ ]:
esox_lucius.plot_species_occurrence()

### One with the smallest area

In [ ]:
non_extinct_fish.head(1).binomial

See all the records(areas) for this binomial

In [ ]:
non_extinct_fish[non_extinct_fish.binomial=="Astatotilapia burtoni"]

In [ ]:
astatotilapia_burtoni = IUCNSpecies(name_species="Astatotilapia burtoni")

In [ ]:
astatotilapia_burtoni.set_data(non_extinct_fish[non_extinct_fish.binomial=="Astatotilapia burtoni"])

In [ ]:
astatotilapia_burtoni.save_shapefile("../data/fish/selection/astatotilapia_burtoni")

In [ ]:
astatotilapia_burtoni.plot_species_occurrence()

This below is just another way of saving the data, i.e., not in a shapefile (which produces multiple files in a separate folder), but a "pickled" data in a single file, compressed. It is usually a lot faster than saving as a shape file, when there is big amounts of data to be saved.

In [ ]:
astatotilapia_burtoni.save_data(dir_name="../data/fish/selection/", method="pickle")

In [ ]:
esox_lucius.save_data(dir_name="../data/fish/selection/", method="pickle")

Since the records are individual shapes, and there could be multiple per species (binomial), let's group them
by the species name, and apply a sum on the shape_area column. This will give us the total area (sum of all polygons) per-species. We want this in order to decide what our "average" species selection will be, for testing the workflow.

In [ ]:
species_area_sum = non_extinct_fish.groupby('binomial')['shape_area'].apply(np.sum)

In [ ]:
species_area_sum.sort(inplace=True)

In [ ]:
species_area_sum.head(10)

In [ ]:
species_area_sum.tail(10)

In [ ]:
species_area_sum.hist(bins=100)

The distribution of areas (in km^2?) is exponential. The median value is rather low:

In [ ]:
species_area_sum.median()

In [ ]:
np.average(species_area_sum)

In [ ]:
species_area_sum[(species_area_sum>45) & (species_area_sum<46)]

### This below is an interesting "average case" with 3 regions, two of which are rather small. It also has a region extending all the way into sea/ocean water

In [ ]:
non_extinct_fish[non_extinct_fish.binomial=="Acrocheilus alutaceus"]

In [ ]:
acrocheilus_alutaceus = IUCNSpecies(name_species="Acrocheilus alutaceus")

In [ ]:
acrocheilus_alutaceus.set_data(non_extinct_fish[non_extinct_fish.binomial=="Acrocheilus alutaceus"])

In [ ]:
acrocheilus_alutaceus.plot_species_occurrence()

In [ ]:
acrocheilus_alutaceus.save_data(dir_name="../data/fish/selection/", method="pickle")

In [ ]:
acrocheilus_alutaceus.save_shapefile("../data/fish/selection/acrocheilus_alutaceus")

The individual polygons are below. They are also "not valid" geometries, makes them interesting for case study.

In [ ]:
non_extinct_fish.loc[10322].geometry

In [ ]:
non_extinct_fish.loc[10320].geometry

In [ ]:
non_extinct_fish.loc[10321].geometry

In [ ]:
non_extinct_binomials = non_extinct_fish.binomial.unique().tolist()

Just persist on storage the list of non-extinct binomials we got from IUCN, so we don't have to filter out again.

In [ ]:
import pickle
pickle.dump(non_extinct_binomials, open("../data/fish/selection/non_extinct_binomials.pkl","wb"))

In [ ]:
non_extinct_binomials.index("Perca fluviatilis")

Let's read Aafke's Excel with lab test data on thermal tolerance for species

In [ ]:
thermal_tolerance_df = pd.read_excel("../data/fish/selection/Lethal_temperature_freshwater_fish_July2016.xlsx")
thermal_tolerance_df.head(10)

In [ ]:
thermal_tolerance_df.columns

In [ ]:
lab_data_list = thermal_tolerance_df['Species '].tolist()

In [ ]:
species_with_lat_lon_after_1960 = year_or_eventdate_1960['species'].unique().tolist()

For every species that is in the lab data list, select it if it is also in the IUCN list of non-extinct binomials, AND in the GBIF filtered-out list of species with latitude/longitude information and event year > 1960:

In [ ]:
filtered_list = []
for species_name in lab_data_list:
    if species_name.strip() in non_extinct_binomials and species_name.strip() in species_with_lat_lon_after_1960:
        filtered_list.append(species_name.strip()) # strip() is just to remove spaces in names

Now get the number of point records for this list of lab species (cross-checked with filtered GBIF records)

In [ ]:
year_or_eventdate_1960[year_or_eventdate_1960.species.isin(filtered_list)].groupby('species')['species'].apply(lambda x: x.count())

In [ ]:
tmp_series = year_or_eventdate_1960[year_or_eventdate_1960.species.isin(filtered_list)].groupby('species')['species'].apply(lambda x: x.count())

In [ ]:
tmp_series[tmp_series>100] # species with more than 50 records

So this is our potential list to choose a representative example from

In [ ]:
tmp_series[tmp_series>100].keys()

In [ ]:
len(tmp_series[tmp_series>100].keys())

In [ ]:
len(filtered_list) # so 20 records are out of the original list because of not enough point-records

In [ ]:
tmp_series[tmp_series>100].sort_values()

Get the GBIF/IUCN data on the first and the last species in the sorted list. That would be our representative choise for testing workflow.

In [ ]:
lepidomeda_mollispinis = IUCNSpecies(name_species="Lepidomeda mollispinis")
lepidomeda_mollispinis.set_data(non_extinct_fish[non_extinct_fish.binomial=="Lepidomeda mollispinis"])
lepidomeda_mollispinis.save_shapefile("../data/fish/selection/lepidomeda_mollispinis")
lepidomeda_mollispinis.plot_species_occurrence()

my_species = GBIFSpecies(name_species="Lepidomeda mollispinis")
# download point-records on these species from: https://drive.google.com/open?id=0B9cazFzBtPuCMGlLMnQzb3hsQm8
data = my_species.load_data("../data/fish/selection/Lepidomeda mollispinis2361895.pkl")
my_species.plot_species_occurrence()

In [ ]:
salmo_trutta = IUCNSpecies(name_species="Salmo trutta")
salmo_trutta.set_data(non_extinct_fish[non_extinct_fish.binomial=="Salmo trutta"])
salmo_trutta.save_shapefile("../data/fish/selection/salmo_trutta")
salmo_trutta.plot_species_occurrence()

my_species = GBIFSpecies(name_species="Salmo trutta")
# download point-records on these species from: https://drive.google.com/open?id=0B9cazFzBtPuCWEIySU5oU2ZfT2M 
data = my_species.load_data("../data/fish/selection/Salmo trutta8215487.pkl")
my_species.plot_species_occurrence() # warning: a lot of point-records, may take a while!!

Interesting, lots of point-records outside the expert range? in particular in the US.

###  Let's overlay the point-records with the range-maps. This will filter-out only the point records that belong to the rangemap area.

my_species contains the GBIF records, salmo_trutta contains the geometries.

In [ ]:
my_species.overlay(salmo_trutta)

In [ ]:
my_species.plot_species_occurrence()